### PR TITLE
server: fix volume migration on user vm scale

### LINKED ISSUE #6701: No permission to migrate volume, only root admin can migrate a volume
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
ALL
~~~

##### CLOUDSTACK VERSION

~~~
4.17.0.1
~~~

##### CONFIGURATION

N/A

##### OS / ENVIRONMENT

CentOS 7

##### SUMMARY

When a non-root user try to change the compute offering on a VM he faces the error: " No permission to migrate volume, only root admin can migrate a volume ".

##### STEPS TO REPRODUCE

~~~
Change the compute offering of a powered-off VM
~~~

##### EXPECTED RESULTS

~~~
The compute offering is changed.
~~~

##### ACTUAL RESULTS

~~~
Non-root user faces error " No permission to migrate volume, only root admin can migrate a volume "
~~~

### PR DESCRIPTION
**Pull request by shwstppr**
### Description

Fixes #6701

When volume migration is initiated by system, account check is not needed.

Introduces a new global setting - `allow.diskOffering.change.during.scale.vm`
Determines whether to allow or disallow disk offering change for root volume during scaling of a stopped or running VM. Default value is `true`

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [x] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test

**Comment by shwstppr** [comment]
Moving this to draft, need to revisit this later. Ideally, the normal user should not be allowed to scale VM or change volume offering with volume migration.

**Comment by DaanHoogland** [comment]
> Moving this to draft, need to revisit this later. Ideally, the normal user should not be allowed to scale VM or change volume offering with volume migration.

is ther some technical reason not to allow that @shwstppr ? because from a business point of view I would want them to be able to scale VMs and volumes.

**Comment by shwstppr** [comment]
@DaanHoogland didn't understand you.
I was saying scale VM operation shouldn't be allowed *with migration of volume* and without migration it should be fine.
This was based on the fact that by default user role is not allowed to use `migrateVolume` API.

**Comment by DaanHoogland** [comment]
> @DaanHoogland didn't understand you. I was saying scale VM operation shouldn't be allowed _with migration of volume_ and without migration it should be fine. This was based on the fact that by default user role is not allowed to use `migrateVolume` API.

ok, makes sense, migration not being allowed. sorry I misunderstood.

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by harikrishna-patnala** [comment]
Added a new global setting for scale VM operation "allow.diskOffering.change.during.scale.vm" which determines whether to allow or disallow disk offering change for root volume during scaling of a stopped or running VM. By default it is false, which means disk offering change will not happen during scale VM. This is to keep the old behavior as is.

**Comment by harikrishna-patnala** [comment]
@blueorangutan test matrix

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by harikrishna-patnala** [comment]
@blueorangutan package

**Comment by harikrishna-patnala** [comment]
@blueorangutan test matrix

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 1]
```diff
@@ -1823,6 +1823,26 @@ private Volume changeDiskOfferingForVolumeInternal(VolumeVO volume, Long newDisk
...
+     * @param a
+     * @param b
+     * @return true if a and b are equal excluding 0 and null values.
+     */
+    private boolean compareEqualsIncludingNullOrZero(Long a, Long b) {
+        if ((a != null && a == 0 && b == null) || (a == null && b != null && b == 0)) {
+            return true;
+        }
+
+        if (a == b) {
```
Comment (DaanHoogland) [inline server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:1839]:
```suggestion
        if (a.equals(b)) {
```

**Comment by DaanHoogland** [comment]
> Moving this to draft, need to revisit this later. Ideally, the normal user should not be allowed to scale VM or change volume offering with volume migration.

@shwstppr , what if a provider has zones in different legal entities (USA vs EU)? they would want to allow a user to move data from one to the other.

**Comment by shwstppr** [comment]
@DaanHoogland migration behaviour can now be tweaked with new global setting added by @harikrishna-patnala

**Review by rohityadavcloud — APPROVED** [review]
I haven't tested it but LGTM

**Comment by rohityadavcloud** [comment]
@shwstppr @DaanHoogland @harikrishna-patnala any update on this, is this ready for review/testing? cc @NuxRo

**Comment by harikrishna-patnala** [comment]
This PR is all good except for the failure of the new smoke test on simulator and xenserver. I'll fix that and then this will be ready for merge. @rohityadavcloud

**Comment by DaanHoogland** [comment]
@harikrishna-patnala 

> This PR is all good except for the failure of the new smoke test on simulator and xenserver. I'll fix that and then this will be ready for merge. @rohityadavcloud

it fails on simulator as well (in travis)
```
=== TestName: test_05_scale_vm_dont_allow_disk_offering_change | Status : FAILED ===
```

**Comment by soreana** [comment]
@shwstppr Can you give an update about this PR, please? What is the current status? I'm willing to help if you want.

**Comment by shwstppr** [comment]
@soreana Hi, PR should be good for testing. There are a couple of failures for Xenserver and Simulator which I feel are more related to the test itself and not the backend functionality. I'll be adding a fix for them later today.

**Comment by soreana** [comment]
@shwstppr That sounds great, thank you :)

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test matrix

**Comment by shwstppr** [comment]
@blueorangutan test matrix

**Comment by soreana** [comment]
I've tested this pr, wiith this patch, I can scale the VM as a domain admin.

**Comment by rohityadavcloud** [comment]
Test LGTM
@borisstoyanov @vladimirpetrov can you help test this?

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 2]
```diff
@@ -1823,6 +1823,26 @@ private Volume changeDiskOfferingForVolumeInternal(VolumeVO volume, Long newDisk
...
+            return true;
+        }
+
+        return false;
+    }
```
Comment (GutoVeronezi) [inline server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:1844]: We could moved this method to `NumbersUtils` and add unit tests.

As zero and null will be considered the same, we can simplify the code:

`org.apache.commons.lang3.ObjectUtils`

```suggestion
    private boolean compareEqualsIncludingNullOrZero(Long a, Long b) {
        a = ObjectUtils.defaultIfNull(a, 0L);
        b = ObjectUtils.defaultIfNull(b, 0L);

        return a.equals(b);
    }
```

**Comment by rohityadavcloud** [comment]
@blueorangutan package

**Comment by rohityadavcloud** [comment]
env issue - pkg is okay for EL7 and others (will rekick after smoketests)
@blueorangutan test

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 3]
```diff
@@ -1823,6 +1824,21 @@ private Volume changeDiskOfferingForVolumeInternal(VolumeVO volume, Long newDisk
...
+     * @param a
+     * @param b
+     * @return true if a and b are equal excluding 0 and null values.
+     */
+    private boolean compareEqualsIncludingNullOrZero(Long a, Long b) {
+        a = ObjectUtils.defaultIfNull(a, 0L);
+        b = ObjectUtils.defaultIfNull(b, 0L);
+
+        return a.equals(b);
+    }
```
Comment (GutoVeronezi) [inline server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:1840]: @shwstppr, we could move this method to `com.cloud.utils.NumbersUtils` and add some unit tests.

File: server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java [hunk 4]
```diff
@@ -1823,6 +1824,21 @@ private Volume changeDiskOfferingForVolumeInternal(VolumeVO volume, Long newDisk
...
 
+    /**
+     * This method is to compare long values, in miniops and maxiops a or b can be null or 0.
+     * Use this method to treat 0 and null as same
+     *
+     * @param a
+     * @param b
+     * @return true if a and b are equal excluding 0 and null values.
+     */
+    private boolean compareEqualsIncludingNullOrZero(Long a, Long b) {
```
Comment (GutoVeronezi) [inline server/src/main/java/com/cloud/storage/VolumeApiServiceImpl.java:1835]: @shwstppr, what do you think about renaming the method to `compareLongsDefaultingNullToZero`?

File: server/src/main/java/com/cloud/vm/UserVmManager.java [hunk 5]
```diff
@@ -47,9 +47,12 @@
  */
 public interface UserVmManager extends UserVmService {
     String EnableDynamicallyScaleVmCK = "enable.dynamic.scale.vm";
+    String AllowDiskOfferingChangeDuringScaleVmCK = "allow.diskOffering.change.during.scale.vm";
     String AllowUserExpungeRecoverVmCK ="allow.user.expunge.recover.vm";
     ConfigKey<Boolean> EnableDynamicallyScaleVm = new ConfigKey<Boolean>("Advanced", Boolean.class, EnableDynamicallyScaleVmCK, "false",
         "Enables/Disables dynamically scaling a vm", true, ConfigKey.Scope.Zone);
+    ConfigKey<Boolean> AllowDiskOfferingChangeDuringScaleVm = new ConfigKey<Boolean>("Advanced", Boolean.class, AllowDiskOfferingChangeDuringScaleVmCK, "false",
+            "Determines whether to allow or disallow disk offering change for root volume during scaling of a stopped or running vm", true, ConfigKey.Scope.Zone);
```
Comment (GutoVeronezi) [inline server/src/main/java/com/cloud/vm/UserVmManager.java:55]: I think we could remove the `String` variable:
```suggestion
    String AllowUserExpungeRecoverVmCK ="allow.user.expunge.recover.vm";
    ConfigKey<Boolean> EnableDynamicallyScaleVm = new ConfigKey<Boolean>("Advanced", Boolean.class, EnableDynamicallyScaleVmCK, "false",
        "Enables/Disables dynamically scaling a vm", true, ConfigKey.Scope.Zone);
    ConfigKey<Boolean> AllowDiskOfferingChangeDuringScaleVm = new ConfigKey<Boolean>("Advanced", Boolean.class, "allow.diskOffering.change.during.scale.vm", "false",
            "Determines whether to allow or disallow disk offering change for root volume during scaling of a stopped or running vm", true, ConfigKey.Scope.Zone);
```

File: server/src/main/java/com/cloud/vm/UserVmManagerImpl.java [hunk 6]
```diff
@@ -2036,7 +2036,14 @@ private void validateDiskOfferingChecks(ServiceOfferingVO currentServiceOffering
         }
     }
 
-    private void changeDiskOfferingForRootVolume(Long vmId, DiskOfferingVO newDiskOffering, Map<String, String> customParameters) throws ResourceAllocationException {
+    private void changeDiskOfferingForRootVolume(Long vmId, DiskOfferingVO newDiskOffering, Map<String, String> customParameters, Long zoneId) throws ResourceAllocationException {
+
+        if (!AllowDiskOfferingChangeDuringScaleVm.valueIn(zoneId)) {
+            if (s_logger.isDebugEnabled()) {
+                s_logger.debug(String.format("Disk Offering change for the root volume is disabled during the compute offering change operation. Please check the setting %s", AllowDiskOfferingChangeDuringScaleVm.key()));
```
Comment (GutoVeronezi) [inline server/src/main/java/com/cloud/vm/UserVmManagerImpl.java:2043]:
```suggestion
                s_logger.debug(String.format("Changing the disk offering of the root volume during the compute offering change operation is disabled. Please check the setting [%s].", AllowDiskOfferingChangeDuringScaleVm.key()));
```

File: server/src/main/java/com/cloud/vm/UserVmManager.java [hunk 7]
```diff
@@ -47,9 +47,12 @@
  */
 public interface UserVmManager extends UserVmService {
     String EnableDynamicallyScaleVmCK = "enable.dynamic.scale.vm";
+    String AllowDiskOfferingChangeDuringScaleVmCK = "allow.diskOffering.change.during.scale.vm";
```
Comment (rohityadavcloud) [inline server/src/main/java/com/cloud/vm/UserVmManager.java:50]:
```suggestion
    String AllowDiskOfferingChangeDuringScaleVmCK = "allow.diskoffering.change.during.scale.vm";
```

**Comment by rohityadavcloud** [comment]
I've committed suggestions, there's one outstanding on refactoring to a utils which can be done in a separate PR as #shwstppr is on holiday this week. I'll merge this to keep effort on-going for 4.17.2.0 release.
@blueorangutan package

**Pull request merged**